File: src/HomeGuard.Client/wwwroot/js/homeguard-db.js
// homeguard-db.js
// IndexedDB wrapper for the Blazor client.
// Exposes a simple key-value store and an Outbox queue.
// Called via IJSRuntime from HomeGuardDb.cs.

const DB_NAME    = 'HomeGuard';
const DB_VERSION = 2;

let _db = null;

// A byte[] argument crossing the .NET->JS interop boundary doesn't consistently arrive
// as one JS type. Confirmed live (not just Uint8Array as first assumed): a byte[] that
// isn't statically typed as byte[] at the JS interop call site — true here, `data` is
// just one field of an anonymous object — arrives wrapped in Blazor's own interop wire
// format, {"__byte[]": "<base64>"}, not a raw Uint8Array. IndexedDB's structured-clone
// storage preserves whichever shape it was handed, so a value read back later can be
// any of them; this is the one place they all get reconciled to the base64 string
// HomeGuardDb.cs's GetBytesFromBase64() actually needs.
function toBase64(data) {
    if (typeof data === 'string') return data;
    if (data instanceof ArrayBuffer) data = new Uint8Array(data);
    if (data instanceof Uint8Array) {
        let binary = '';
        for (let i = 0; i < data.length; i++) binary += String.fromCharCode(data[i]);
        return btoa(binary);
    }
    if (data && typeof data === 'object' && typeof data['__byte[]'] === 'string') {
        return data['__byte[]'];
    }
    throw new Error('Unrecognized blob data shape: ' + Object.prototype.toString.call(data));
}

async function openDb() {
    if (_db) return _db;

    return new Promise((resolve, reject) => {
        const req = indexedDB.open(DB_NAME, DB_VERSION);

        req.onupgradeneeded = e => {
            const db = e.target.result;

            // Outbox: pending operations to sync to the server.
            if (!db.objectStoreNames.contains('outbox')) {
                const outbox = db.createObjectStore('outbox', { keyPath: 'clientOperationId' });
                outbox.createIndex('by_status',    'status',    { unique: false });
                outbox.createIndex('by_createdAt', 'createdAt', { unique: false });
            }

            // Cache: local copies of server data for offline reads.
            if (!db.objectStoreNames.contains('cache')) {
                db.createObjectStore('cache', { keyPath: 'key' });
            }

            // Blob outbox: file uploads pending a server round-trip. Kept separate from
            // 'outbox' — that store's payloadJson goes through the batch JSON endpoint,
            // which is the wrong shape for a file that can be several MB.
            if (!db.objectStoreNames.contains('blobOutbox')) {
                const blobOutbox = db.createObjectStore('blobOutbox', { keyPath: 'clientOperationId' });
                blobOutbox.createIndex('by_createdAt', 'createdAt', { unique: false });
            }
        };

        req.onsuccess = e => { _db = e.target.result; resolve(_db); };
        req.onerror   = e => reject(e.target.error);
    });
}

// ── Outbox ────────────────────────────────────────────────────────────────────

window.homeGuardDb = {

    // Add an entry to the outbox. Status: 'pending' | 'delivering' | 'delivered' | 'failed'
    async outboxAdd(entry) {
        const db = await openDb();
        return new Promise((resolve, reject) => {
            const tx    = db.transaction('outbox', 'readwrite');
            const store = tx.objectStore('outbox');
            const req   = store.put({ ...entry, status: 'pending' });
            req.onsuccess = () => resolve(req.result);
            req.onerror   = () => reject(req.error);
        });
    },

    // Get all pending entries ordered by createdAt.
    async outboxGetPending() {
        const db = await openDb();
        return new Promise((resolve, reject) => {
            const tx      = db.transaction('outbox', 'readonly');
            const index   = tx.objectStore('outbox').index('by_createdAt');
            const results = [];
            const req     = index.openCursor();
            req.onsuccess = e => {
                const cursor = e.target.result;
                if (cursor) {
                    if (cursor.value.status === 'pending' || cursor.value.status === 'failed')
                        results.push(cursor.value);
                    cursor.continue();
                } else {
                    resolve(results);
                }
            };
            req.onerror = () => reject(req.error);
        });
    },

    // Mark entries as delivered and remove them from the outbox.
    async outboxMarkDelivered(clientOperationIds) {
        const db = await openDb();
        return new Promise((resolve, reject) => {
            const tx    = db.transaction('outbox', 'readwrite');
            const store = tx.objectStore('outbox');
            let done = 0;
            for (const id of clientOperationIds) {
                const req = store.delete(id);
                req.onsuccess = () => { if (++done === clientOperationIds.length) resolve(); };
                req.onerror   = () => reject(req.error);
            }
            if (clientOperationIds.length === 0) resolve();
        });
    },

    // Mark an entry as failed (will be retried next sync).
    async outboxMarkFailed(clientOperationId) {
        const db = await openDb();
        return new Promise((resolve, reject) => {
            const tx    = db.transaction('outbox', 'readwrite');
            const store = tx.objectStore('outbox');
            const get   = store.get(clientOperationId);
            get.onsuccess = () => {
                if (!get.result) return resolve();
                const entry = { ...get.result, status: 'failed' };
                const put   = store.put(entry);
                put.onsuccess = () => resolve();
                put.onerror   = () => reject(put.error);
            };
            get.onerror = () => reject(get.error);
        });
    },

    async outboxCount() {
        const db = await openDb();
        return new Promise((resolve, reject) => {
            const tx  = db.transaction('outbox', 'readonly');
            const req = tx.objectStore('outbox').count();
            req.onsuccess = () => resolve(req.result);
            req.onerror   = () => reject(req.error);
        });
    },

    // ── Blob outbox ──────────────────────────────────────────────────────────────

    async blobOutboxAdd(entry) {
        const db = await openDb();
        // .NET's own byte[]-argument marshalling picks its own wire format for `data`
        // (a raw Uint8Array in some paths, a base64 string in others) — normalizing to
        // one shape here means blobOutboxGetPending never has to guess which one it's
        // reading back.
        const normalized = { ...entry, data: toBase64(entry.data) };
        return new Promise((resolve, reject) => {
            const tx    = db.transaction('blobOutbox', 'readwrite');
            const store = tx.objectStore('blobOutbox');
            const req   = store.put(normalized);
            req.onsuccess = () => resolve(req.result);
            req.onerror   = () => reject(req.error);
        });
    },

    // Get all pending blob uploads ordered by createdAt. `data` is always returned as a
    // base64 string — HomeGuardDb.cs's GetBytesFromBase64() depends on that — even for
    // rows written before blobOutboxAdd started normalizing on the way in. A row whose
    // `data` isn't convertible at all (corrupt beyond recovery) is dropped and deleted
    // instead of taking the whole flush down with it — see the CS2012-adjacent crash
    // this was written for: one bad queued capture blocked every future sync attempt.
    async blobOutboxGetPending() {
        const db = await openDb();
        return new Promise((resolve, reject) => {
            const tx      = db.transaction('blobOutbox', 'readonly');
            const index   = tx.objectStore('blobOutbox').index('by_createdAt');
            const results = [];
            const toDrop  = [];
            const req     = index.openCursor();
            req.onsuccess = e => {
                const cursor = e.target.result;
                if (cursor) {
                    try {
                        results.push({ ...cursor.value, data: toBase64(cursor.value.data) });
                    } catch (err) {
                        console.warn('[homeguard-db] dropping unreadable blobOutbox entry',
                            cursor.value.clientOperationId, err);
                        toDrop.push(cursor.value.clientOperationId);
                    }
                    cursor.continue();
                } else {
                    Promise.all(toDrop.map(id => homeGuardDb.blobOutboxRemove(id)))
                        .catch(() => { /* best-effort cleanup — a failed delete just retries next time */ })
                        .then(() => resolve(results));
                }
            };
            req.onerror = () => reject(req.error);
        });
    },

    async blobOutboxRemove(clientOperationId) {
        const db = await openDb();
        return new Promise((resolve, reject) => {
            const tx  = db.transaction('blobOutbox', 'readwrite');
            const req = tx.objectStore('blobOutbox').delete(clientOperationId);
            req.onsuccess = () => resolve();
            req.onerror   = () => reject(req.error);
        });
    },

    async blobOutboxCount() {
        const db = await openDb();
        return new Promise((resolve, reject) => {
            const tx  = db.transaction('blobOutbox', 'readonly');
            const req = tx.objectStore('blobOutbox').count();
            req.onsuccess = () => resolve(req.result);
            req.onerror   = () => reject(req.error);
        });
    },

    // ── Cache ─────────────────────────────────────────────────────────────────

    async cacheSet(key, valueJson) {
        const db = await openDb();
        return new Promise((resolve, reject) => {
            const tx  = db.transaction('cache', 'readwrite');
            const req = tx.objectStore('cache').put({ key, value: valueJson, savedAt: Date.now() });
            req.onsuccess = () => resolve();
            req.onerror   = () => reject(req.error);
        });
    },

    async cacheGet(key) {
        const db = await openDb();
        return new Promise((resolve, reject) => {
            const tx  = db.transaction('cache', 'readonly');
            const req = tx.objectStore('cache').get(key);
            req.onsuccess = () => resolve(req.result ? req.result.value : null);
            req.onerror   = () => reject(req.error);
        });
    },

    async cacheDelete(key) {
        const db = await openDb();
        return new Promise((resolve, reject) => {
            const tx  = db.transaction('cache', 'readwrite');
            const req = tx.objectStore('cache').delete(key);
            req.onsuccess = () => resolve();
            req.onerror   = () => reject(req.error);
        });
    },
};

File: src/HomeGuard.Client/Services/HomeGuardDb.cs
using Microsoft.Extensions.Logging;
using Microsoft.JSInterop;
using System.Text.Json;

namespace HomeGuard.Client.Services;

/// <summary>
/// C# wrapper over the homeguard-db.js IndexedDB API.
/// All methods are async and call into JS via IJSRuntime.
/// </summary>
public sealed class HomeGuardDb
{
    private readonly IJSRuntime _js;
    private readonly ILogger<HomeGuardDb> _logger;

    public HomeGuardDb(IJSRuntime js, ILogger<HomeGuardDb> logger)
    {
        _js     = js;
        _logger = logger;
    }

    // ── Outbox ────────────────────────────────────────────────────────────────

    public async Task OutboxAddAsync(OutboxEntryLocal entry)
    {
        var obj = new
        {
            clientOperationId = entry.ClientOperationId,
            operationType     = entry.OperationType,
            payloadJson       = entry.PayloadJson,
            createdAt         = entry.CreatedAt.ToUnixTimeMilliseconds(),
        };
        await _js.InvokeVoidAsync("homeGuardDb.outboxAdd", obj);
    }

    public async Task<IReadOnlyList<OutboxEntryLocal>> OutboxGetPendingAsync()
    {
        var raw = await _js.InvokeAsync<JsonElement[]>("homeGuardDb.outboxGetPending");
        return raw.Select(e => new OutboxEntryLocal(
            ClientOperationId: e.GetProperty("clientOperationId").GetString()!,
            OperationType:     e.GetProperty("operationType").GetString()!,
            PayloadJson:       e.GetProperty("payloadJson").GetString()!,
            CreatedAt: DateTimeOffset.FromUnixTimeMilliseconds(
                           e.GetProperty("createdAt").GetInt64())
        )).ToList();
    }

    public async Task OutboxMarkDeliveredAsync(IEnumerable<string> clientOperationIds)
        => await _js.InvokeVoidAsync("homeGuardDb.outboxMarkDelivered",
               clientOperationIds.ToArray());

    public async Task OutboxMarkFailedAsync(string clientOperationId)
        => await _js.InvokeVoidAsync("homeGuardDb.outboxMarkFailed", clientOperationId);

    public async Task<int> OutboxCountAsync()
        => await _js.InvokeAsync<int>("homeGuardDb.outboxCount");

    // ── Blob outbox ──────────────────────────────────────────────────────────────
    // Separate from the JSON outbox above: a file can be several MB, and the batch
    // endpoint (api/sync/batch) is sized for small command payloads, not attachments.
    // Each pending upload gets its own POST to api/blobs/upload when flushed.

    public async Task BlobOutboxAddAsync(PendingBlobUpload entry)
    {
        var obj = new
        {
            clientOperationId = entry.ClientOperationId,
            // Confirmed live, the actual root cause: a byte[]-typed value nested in a
            // generic interop argument gets Blazor's own byte-array-transfer treatment —
            // the JSON payload carries only a transient reference id ({"__byte[]": 0}),
            // valid for the duration of this one call, not the real bytes. IndexedDB
            // storage outlives that call by definition, so anything read back later was
            // always going to be an unrecoverable dangling reference. Converting to a
            // plain string here means the property is never byte[]-typed at the interop
            // boundary at all, so that whole path never triggers.
            data              = Convert.ToBase64String(entry.Data),
            mimeType          = entry.MimeType,
            fileName          = entry.FileName,
            ownerEntityId     = entry.OwnerEntityId.ToString(),
            ownerEntityType   = entry.OwnerEntityType,
            createdAt         = entry.CreatedAt.ToUnixTimeMilliseconds(),
        };
        await _js.InvokeVoidAsync("homeGuardDb.blobOutboxAdd", obj);
    }

    /// <summary>
    /// homeguard-db.js's own blobOutboxGetPending already normalizes `data` to a base64
    /// string, but this doesn't trust that that fix has actually reached the browser
    /// being read from (a PWA's cached JS can lag a fresh publish) — it accepts
    /// whichever shape .NET's own byte[]-argument interop marshalling happened to write
    /// under the old, unnormalized path: a plain base64 string, a JSON array of byte
    /// values, or Blazor's own {"__byte[]": ...} interop wire format for a byte[] that
    /// wasn't statically typed as byte[] at the call site — confirmed live to carry
    /// either the base64 payload directly, *or*, when the value is a transient
    /// byte-array-transfer reference id (a plain number) rather than the payload, no
    /// recoverable data at all: that reference is only valid for the one interop call
    /// that created it, which IndexedDB storage necessarily outlives. Root-caused and
    /// fixed at the write side now (BlobOutboxAddAsync converts to a plain string before
    /// the interop call, so new rows never take this path) — this stays to explain, and
    /// cleanly drop, whatever is already stuck in an existing IndexedDB from before that.
    /// </summary>
    private static byte[] ReadBlobData(JsonElement entry)
    {
        var data = entry.GetProperty("data");
        if (data.ValueKind == JsonValueKind.Object && data.TryGetProperty("__byte[]", out var wrapped))
        {
            if (wrapped.ValueKind == JsonValueKind.String)
                return wrapped.GetBytesFromBase64();

            throw new InvalidOperationException(
                $"'data' is a transient byte-array interop reference ({wrapped.GetRawText()}), " +
                "not the actual bytes — this row's real data no longer exists anywhere to recover.");
        }

        return data.ValueKind switch
        {
            JsonValueKind.String => data.GetBytesFromBase64(),
            JsonValueKind.Array  => [.. data.EnumerateArray().Select(x => (byte)x.GetInt32())],
            JsonValueKind.Object => [.. data.EnumerateObject()
                                          .OrderBy(p => int.Parse(p.Name, System.Globalization.CultureInfo.InvariantCulture))
                                          .Select(p => (byte)p.Value.GetInt32())],
            var kind => throw new InvalidOperationException($"Unexpected blob data shape: {kind}"),
        };
    }

    public async Task<IReadOnlyList<PendingBlobUpload>> BlobOutboxGetPendingAsync()
    {
        var raw    = await _js.InvokeAsync<JsonElement[]>("homeGuardDb.blobOutboxGetPending");
        var result = new List<PendingBlobUpload>(raw.Length);

        foreach (var e in raw)
        {
            string? clientOperationId = null;
            try
            {
                clientOperationId = e.GetProperty("clientOperationId").GetString();
                result.Add(new PendingBlobUpload(
                    ClientOperationId: clientOperationId!,
                    Data:              ReadBlobData(e),
                    MimeType:          e.GetProperty("mimeType").GetString()!,
                    FileName:          e.GetProperty("fileName").GetString()!,
                    OwnerEntityId:     Guid.Parse(e.GetProperty("ownerEntityId").GetString()!),
                    OwnerEntityType:   e.GetProperty("ownerEntityType").GetString()!,
                    CreatedAt: DateTimeOffset.FromUnixTimeMilliseconds(
                                   e.GetProperty("createdAt").GetInt64())
                ));
            }
            catch (Exception ex)
            {
                // Two guesses at the actual wire shape have both been wrong — dump the
                // real thing instead of a third guess. Capped well short of the full
                // payload (a photo's base64 alone can run six figures of characters);
                // enough to show the property names/shape, not the file itself.
                string dataShape;
                try
                {
                    var text = e.TryGetProperty("data", out var data) ? data.GetRawText() : "<no data property>";
                    dataShape = text.Length > 200 ? text[..200] + "…" : text;
                }
                catch (Exception shapeEx)
                {
                    dataShape = $"<failed to read shape: {shapeEx.Message}>";
                }

                // One row this defensive a parse still can't make sense of is genuinely
                // corrupt — drop it (and remove it, if its id was even readable) instead
                // of taking every other queued upload down with it.
                _logger.LogWarning(ex, "Dropping unreadable blobOutbox entry {ClientOperationId} — data shape: {DataShape}",
                    clientOperationId, dataShape);
                if (clientOperationId is not null)
                    await BlobOutboxRemoveAsync(clientOperationId);
            }
        }

        return result;
    }

    public async Task BlobOutboxRemoveAsync(string clientOperationId)
        => await _js.InvokeVoidAsync("homeGuardDb.blobOutboxRemove", clientOperationId);

    public async Task<int> BlobOutboxCountAsync()
        => await _js.InvokeAsync<int>("homeGuardDb.blobOutboxCount");

    // ── Cache ─────────────────────────────────────────────────────────────────

    public async Task CacheSetAsync<T>(string key, T value)
    {
        var json = JsonSerializer.Serialize(value, Json.Options);
        await _js.InvokeVoidAsync("homeGuardDb.cacheSet", key, json);
    }

    public async Task<T?> CacheGetAsync<T>(string key)
    {
        var json = await _js.InvokeAsync<string?>("homeGuardDb.cacheGet", key);
        return json is null ? default : JsonSerializer.Deserialize<T>(json, Json.Options);
    }

    public Task CacheDeleteAsync(string key)
        => _js.InvokeVoidAsync("homeGuardDb.cacheDelete", key).AsTask();
}

/// <summary>Client-side outbox entry stored in IndexedDB.</summary>
public sealed record OutboxEntryLocal(
    string ClientOperationId,
    string OperationType,
    string PayloadJson,
    DateTimeOffset CreatedAt
);

/// <summary>A file capture waiting for a server round-trip, stored in IndexedDB.</summary>
public sealed record PendingBlobUpload(
    string ClientOperationId,
    byte[] Data,
    string MimeType,
    string FileName,
    Guid OwnerEntityId,
    string OwnerEntityType,
    DateTimeOffset CreatedAt
);
